cEP0002.md

```diff
@@ -0,0 +1,166 @@
+# Next Gen Bear Design
+
+| Metadata     |                                         |
+| ------------ |-----------------------------------------|
+| cEP          | 2                                       |
+| Version      | 0.1                                     |
+| Title        | Next gen bear design                    |
+| Authors      | Udayan Tandon <udayan12167@iiitd.ac.in> |
+| Status       | Proposed                                |
+| Type         | Process                                 |
+
+# Abstract
+
+This document describes how the next generation bears will be designed and
+implemented.
+
+This cEP describes a new implementation of bears in coala which should make
+ them simpler and make their design more intuitive and better.
+
+# Introduction
+
+Currently the bear design is clunky and bad. The current design makes global
+bears dependant on local bears through it's queuing mechanism. It uses
+communication signals over a queue to indicate that global bears are done with
+their execution and local bears can begin. The queueing mechanism doesn't work
+very well as it sometimes has synchronization problems. Also if local bears
+fail to run it stalls the global bears. Also we have a permanent logger thread.
+An extra logger thread consumes time and resources of the system which we can
+avoid.
+
+The new proposal aims to eliminate the use of the queuing control signals.
+Utilising a multiprocessing pool to generate the results from the bears. No
+longer will the local bears be dependent on global bears to finish. The logging
+will also be handled by the main thread itself instead of utilising an extra
+thread.
+
+# Proposed Change
+
+Here is the detailed implementation stepwise:
+
+1. coala will have a core multiprocessing pool.
+2. The `Bear` class will get two new methods. `__execute_task()` and
+`generate_tasks()`.
+    * The `generate_tasks()` is resposible for generating the correct
+    arguments for the upcoming tasks.
+    * The `__execute_task()` method is called in parallel by the
+    multiprocessing pool on as many available threads.
+3. Each bear will load the core coala pool with different types of tasks.
+    * Global bears fire tasks that process all files.
+    * Local bears fire tasks for each individual file.
+6. As the task runs it places it's generated results into a result queue
+during runtime and continues to do so till the task completes.
```

Comment:
when using a pool we need to wait until the task is done^^ the only way I can imagine this is possible using generators and `next`, but you would also need to query each running task for results. This isn't really practicable imo^^